cpp/src/arrow/filesystem/util_internal.cc

```diff
@@ -19,6 +19,9 @@
 
 #include <algorithm>
 #include <cerrno>
+#include <cctype>
+#include <aws/core/utils/HashingUtils.h>
+#include <aws/core/utils/base64/Base64.h>
```

Comment:
Hmm. `util_internal.cc` is shared by not only the S3 filesystem implementation but also all filesystem implementations.
Can we avoid using `util_internal.cc` for S3 specific utility?

Reply:
+1

Reply:
I would try to move it into the s3_internal.h